src/services/asset.service.ts

```diff
@@ -0,0 +1,123 @@
+import { Injectable } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+
+import { Repository } from 'typeorm';
+
+import { AssetEntity } from '@app/database';
+import { filterFalsy, GenericValueEntity } from '@app/utils';
+import { AssetInfo } from '@app/http';
+
+@Injectable()
+export class AssetService {
+    constructor(@InjectRepository(AssetEntity) private readonly _repository: Repository<AssetEntity>) {}
+
+    getByCompositeKey = async (compositeKey: string): Promise<AssetEntity> => {
+        return this._repository.findOne({
+            where: {
+                id: compositeKey,
+            },
+        });
+    };
+
+    getExtra = async (): Promise<{ id: string; extra: [] }[]> => {
+        const extra = this._repository.createQueryBuilder('assets').select(['id', 'extra']);
+        return await extra.getRawMany();
+    };
+
+    createOrUpdateAssetValue = async (compositeKey: string, value: any): Promise<AssetEntity> => {
+        let entity = await this.getByCompositeKey(compositeKey);
+
+        // If entity does not exists, we create with a new one
+        if (!entity) {
+            entity = new AssetEntity({
+                id: compositeKey,
+                value,
+            });
+        } else {
+            // Otherwise, we just update the propertiess
+            entity.id = compositeKey;
+            entity.value = value;
+        }
+
+        await this._repository.save(entity);
+        return entity;
+    };
+
+    chainAssetCreateOrUpdateValue = async (getAssetInfo: AssetInfo[]): Promise<void> => {
+        for (const key of filterFalsy(getAssetInfo)) {
+            if (key) {
+                const compositeKey = `${key.symbol.toLowerCase()}_${Object.keys(key)[0]}`;
+                const value = { [Object.keys(key)[0]]: Object.values(key)[0], last_updated_at: new Date() };
+
+                await this.createOrUpdateAssetValue(compositeKey, value);
+            }
+        }
+    };
+
+    owneAssetCreateOrUpdateValue = async (getAssetInfo: any, name: string): Promise<void> => {
+        for (const key in filterFalsy(getAssetInfo)) {
+            if (key) {
+                const compositeKey = `${name.toLowerCase()}_${key}`;
+
+                const value = { [key]: getAssetInfo[key], last_updated_at: new Date() };
+
+                await this.createOrUpdateAssetValue(compositeKey, value);
+            }
+        }
+    };
+
+    createOrUpdateAssetExtra = async (compositeKey: string): Promise<GenericValueEntity> => {
+        // Helper function to append historical data to the extra column
+        const entity = await this.getByCompositeKey(compositeKey);
+
+        const query = await this._repository.query(`
+            UPDATE assets
+            SET extra = extra || '${JSON.stringify(entity.value)}'::jsonb
+            WHERE id = '${compositeKey}'
+        `);
+
+        return query;
```

Comment:
No way do it without raw SQL?

Reply:
Couldn't find it for typeorm. Also this query shouldn't affect performance.